components/templates/src/global_fns/load_data.rs

```diff
@@ -0,0 +1,346 @@
+extern crate toml;
+extern crate serde_json;
+
+use utils::fs::{read_file, is_path_in_directory, get_file_time};
+
+use crypto_hash::{Algorithm, hex_digest};
+use chrono::{DateTime, Utc};
+use reqwest::{Client, header};
+use url::Url;
+
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+
+
+use csv::Reader;
+use std::collections::HashMap;
+use tera::{GlobalFn, Value, from_value, to_value, Result, Map};
+use std::ops::BitXor;
+
+static GET_DATA_ARGUMENT_ERROR_MESSAGE: &str = "`load_data`: requires EITHER a `path` or `url` argument";
+
+enum ProvidedArgument {
+    URL(Url),
+    PATH(PathBuf)
+}
+
+
+fn get_cache_key(provided_argument: &ProvidedArgument, format: &String) -> String {
+    let content_based_data = match provided_argument {
+        ProvidedArgument::URL(url) => url.clone().into_string(),
+        ProvidedArgument::PATH(path) => {
+            let file_time = get_file_time(&path).expect("get file time");
+            let file_datetime: DateTime<Utc> = DateTime::from(file_time);
+            format!("{}{}", file_datetime.timestamp_millis().to_string(), path.display())
+        }
+    };
+    return hex_digest(Algorithm::MD5, format!("{}{}", format, content_based_data).as_bytes());
+}
+
+
+fn get_data_from_args(content_path: &PathBuf, args: &HashMap<String, Value>) -> Result<ProvidedArgument> {
+    let path_arg = optional_arg!(
+        String,
+        args.get("path"),
+        GET_DATA_ARGUMENT_ERROR_MESSAGE
+    );
+
+    let url_arg = optional_arg!(
+        String,
+        args.get("url"),
+        GET_DATA_ARGUMENT_ERROR_MESSAGE
+    );
+
+    if !path_arg.is_some().bitxor(url_arg.is_some()) {
+        return Err("GET_DATA_ARGUMENT_ERROR_MESSAGE.into()".into());
+    }
+
+    if let Some(path) = path_arg {
+        let full_path = content_path.join(path);
+        if !full_path.exists() {
+            return Err(format!("{} doesn't exist", full_path.display()).into());
+        }
+        return Ok(ProvidedArgument::PATH(full_path));
+    }
+    else if let Some(url) = url_arg {
+        return Url::parse(&url).map(|parsed_url| ProvidedArgument::URL(parsed_url)).map_err(|e| format!("Failed to parse {} as url: {}", url, e).into());
+    }
+
+    return Err(GET_DATA_ARGUMENT_ERROR_MESSAGE.into());
+}
+
+fn read_data_file(base_path: &PathBuf, full_path: PathBuf) -> Result<String> {
+    if !is_path_in_directory(&base_path, &full_path).map_err(|e| format!("Failed to read data file {}: {}", full_path.display(), e))? {
+        return Err(format!("{} is not inside the base site directory {}", full_path.display(), base_path.display()).into());
+    }
+    return read_file(&full_path)
+        .map_err(|e| format!("`load_data`: error {} loading file {}", full_path.to_str().unwrap(), e).into());
+}
+
+fn get_output_format_from_args(args: &HashMap<String, Value>, provided_argument: &ProvidedArgument) -> Result<String> {
```

Comment:
This function should return an enum of the supported types to avoid checking strings later